# Balint Pandas assignment

In [2]:
# import libraries
import numpy as np
import pandas as pd

In [3]:
# Read in the data
df = pd.read_csv(r'C:\Users\BalintStewart\OneDrive - Health Data Research\Desktop\Magnimind\Pandas.14\ksprojects.csv')
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

1- How many rows of data are in the DataFrame?


In [4]:
# Find the numnber of rows in the df, can do this with .shape attribute and select rows
df.shape[0]

65418

2- What are the names and data types of the columns?


In [5]:
col_types = df.dtypes
print(col_types)

id                 int64
name              object
category          object
main_category     object
deadline          object
goal             float64
launched          object
state             object
backers            int64
country           object
usd_pledged      float64
currency          object
pledged          float64
dtype: object


3- Do any of the columns contain null values?

In [6]:
df.isna().sum()

id                 0
name               0
category           0
main_category      0
deadline           0
goal               0
launched           0
state              0
backers            0
country            0
usd_pledged      745
currency           0
pledged            0
dtype: int64

Q3 answer: yes usd_pledged contains 745 missing values. None of the other columns contain missing values

4- Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.

In [7]:
df[df['state'] == 'successful'].sort_values(by= 'usd_pledged', ascending = False).head(10)

id                                               name  \
12253  1033978702             OUYA: A New Kind of Video Game Console   
26917  1152735261     Pono Music - Where Your Soul Rediscovers Music   
45290  1074800880                  Bloodstained: Ritual of the Night   
46419  1045500705                            Zombicide: Black Plague   
59897   132311747                                   Massive Darkness   
27604  1265992004     The Micro: The First Truly Consumer 3D Printer   
45456  1194278206         Yooka-Laylee - A 3D Platformer Rare-vival!   
22674   127494612  Reaper Miniatures Bones II: The Return Of Mr B...   
47232  1194924825  Reaper Miniatures Bones 3: The Search for Mr. ...   
57199  1241205786  OSSIC X: The first 3D audio headphones calibra...   

              category main_category             deadline      goal  \
12253  Gaming Hardware         Games  2012-08-09 07:00:00  950000.0   
26917            Sound    Technology  2014-04-15 19:30:42  800000.0   
45290      Video Games         Games  2015-06-13 05:00:00  500000.0   
46419   Tabletop Games         Games  2015-07-07 02:00:00  125000.0   
59897   Tabletop Games         Games  2016-07-08 01:00:00  200000.0   
27604      3D Printing    Technology  2014-05-07 14:45:46   50000.0   
45456      Video Games         Games  2015-06-17 01:00:00  175000.0   
22674   Tabletop Games         Games  2013-10-27 01:00:00   30000.0   
47232   Tabletop Games         Games  2015-07-26 00:00:00   30000.0   
57199            Sound    Technology  2016-04-21 17:00:00  100000.0   

                  launched       state  backers country  usd_pledged currency  \
12253  2012-07-10 14:44:41  successful    63416      US    8596480.0      USD   
26917  2014-03-11 18:30:42  successful    18220      US    6225360.0      USD   
45290  2015-05-11 18:58:41  successful    64867      US    5545990.0      USD   
46419  2015-06-08 20:59:38  successful    20915      US    4079200.0      USD   
59897  2016-06-07 21:02:31  successful    22361      US    3560640.0      USD   
27604  2014-04-07 14:45:46  successful    11855      US    3401360.0      USD   
45456  2015-05-01 17:57:02  successful    73206      GB    3225230.0      GBP   
22674  2013-10-01 16:00:34  successful    14964      US    3169610.0      USD   
47232  2015-07-07 17:00:43  successful    13465      US    2730370.0      USD   
57199  2016-02-23 09:07:07  successful    10263      US    2708470.0      USD   

         pledged  
12253  8596480.0  
26917  6225360.0  
45290  5545990.0  
46419  4079200.0  
59897  3560640.0  
27604  3401360.0  
45456  2090100.0  
22674  3169610.0  
47232  2730370.0  
57199  2708470.0

5- Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.

In [8]:
df['average_per_backer'] = df['usd_pledged'] / df['backers']
df.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  average_per_backer  
0    40.00           10.000000  
1   700.00           50.000000  
2  1367.00           42.718750  
3  6575.00          273.958333  
4   501.66           22.802727

6- What happened to the rows with 0 backers? How can this be dealt with?


In [9]:
# Select rows with zero backers, look at average_per_backer column
df[df['backers'] == 0].head()

# Check if there are films that had zero backers but non-zero usd_pledged
df[(df['backers'] == 0) & (df['usd_pledged'] != 0)].head()

id                                               name  \
9490   1189637889                              "Natural Born Hooker"   
15111  1244477824  Statpedia - The Collaborative Search Engine fo...   
17484  1084993156  Willy Tea, Chris Doud and Chums: The Children'...   
17532  1379649454  Juiette Z. Payne presents debut EP "Eternal Da...   
27334  1349042579                         O'Films Production Launch!   

           category main_category             deadline     goal  \
9490        Theater       Theater  2012-04-24 08:59:00  25000.0   
15111      Software    Technology  2012-12-14 10:48:14    500.0   
17484         Music         Music  2013-04-09 06:06:11   3750.0   
17532         Music         Music  2013-04-11 15:50:25   5000.0   
27334  Film & Video  Film & Video  2014-04-30 02:58:25   6000.0   

                  launched       state  backers country  usd_pledged currency  \
9490   2012-03-14 19:05:44   undefined        0    N,"0          NaN      USD   
15111  2012-11-14 10:48:14   undefined        0    N,"0          NaN      USD   
17484  2013-03-26 05:06:11  successful        0    N,"0          NaN      USD   
17532  2013-03-02 15:50:25  successful        0    N,"0          NaN      USD   
27334  2014-04-08 02:58:25   undefined        0    N,"0          NaN      USD   

       pledged  average_per_backer  
9490    2621.0                 NaN  
15111    500.0                 NaN  
17484   4593.0                 NaN  
17532   5250.0                 NaN  
27334   6168.0                 NaN

Q6 answer: Films with 0 backers have 'NaN' for average because dividing a zero or another number by zero is undefined or infinite. There are some rows that have 0 backers but non-zero usd_pledged. Instead of removing these rows I would therefore instead replace the zeros in this column to '1' using the following code:

df['backers_replaced'] = df['backers'].replace(0, 1)


7- Drop all rows with 0 backers then repeat the previous exercise.

In [14]:
# Drop all rows with 0 backers (boolean indexing)
df_filtered = df.loc[df['backers'] != 0]

# Check rows have been dropped
df_filtered.shape[0]

# Create a new column with average usd pledged per backer
df_filtered['average_per_backer'] = df_filtered['usd_pledged'] / df_filtered['backers']


C:\Users\BalintStewart\AppData\Local\Temp\ipykernel_3252\1217970074.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['average_per_backer'] = df_filtered['usd_pledged'] / df_filtered['backers']


8- Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.

In [17]:
# Create index and df columns as lists of values
index_labels = ['A','B','C','D','E','F','G','H','I','J']
primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
fibo = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


In [19]:
# use pd.DataFrame to combine these lists into a DF
df = pd.DataFrame({'Prime numbers': primes,
                   'Fibonnacci numbers': fibo},
                   index = index_labels)

df.head()

Prime numbers  Fibonnacci numbers
A              2                   0
B              3                   1
C              5                   1
D              7                   2
E             11                   3

9- Merge these two DataFrames into a single DataFrame: nycflights13.

In [21]:
# Read in the flights and airline data from URL
flights = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv')
airline = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv')



year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00  
2      1089     5      40  2013-01-01 05:00:00  
3      1576     5      45  2013-01-01 05:00:00  
4       762     6       0  2013-01-01 06:00:00

In [ ]:
# Preview the data to find column to merge on
flights.head()

In [22]:
# Preview the data to find column to merge on: need to merge on 'carrier' column
airline.head()

carrier                    name
0      9E       Endeavor Air Inc.
1      AA  American Airlines Inc.
2      AS    Alaska Airlines Inc.
3      B6         JetBlue Airways
4      DL    Delta Air Lines Inc.

In [23]:
# Merge both datasets on carrier column
nycflights13 = flights.merge(airline, on = 'carrier')

# Preview the merged dataset
nycflights13.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour                    name  
0      1400     5      15  2013-01-01 05:00:00   United Air Lines Inc.  
1      1416     5      29  2013-01-01 05:00:00   United Air Lines Inc.  
2      1089     5      40  2013-01-01 05:00:00  American Airlines Inc.  
3      1576     5      45  2013-01-01 05:00:00         JetBlue Airways  
4       762     6       0  2013-01-01 06:00:00    Delta Air Lines Inc.

10- Are there any missing values in this DataFrame? If so, drop them.

In [24]:
# Check for missing values
nycflights13.isna().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
name                 0
dtype: int64

In [27]:
# Drop missing values
nycflights13_dropna = nycflights13.dropna()

# Check NAs have been dropped
nycflights13_dropna.isna().sum()

year              0
month             0
day               0
dep_time          0
sched_dep_time    0
dep_delay         0
arr_time          0
sched_arr_time    0
arr_delay         0
carrier           0
flight            0
tailnum           0
origin            0
dest              0
air_time          0
distance          0
hour              0
minute            0
time_hour         0
name              0
dtype: int64

11- Merge these two DataFrames into a single DataFrame called census.

In [28]:
# Load the two datasets to be merged
populations = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv')
divisions = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv')

In [29]:
# Preview the data to identify column to merge on
populations.head()

state      2010      2011      2012      2013      2014      2015  \
0     Alabama   4785492   4799918   4815960   4829479   4843214   4853875   
1      Alaska    714031    722713    731089    736879    736705    737709   
2     Arizona   6408312   6467163   6549634   6624617   6719993   6817565   
3    Arkansas   2921995   2939493   2950685   2958663   2966912   2977853   
4  California  37332685  37676861  38011074  38335203  38680810  38993940   

       2016  
0   4863300  
1    741894  
2   6931071  
3   2988248  
4  39250017

In [30]:
# Preview the data to identify column to merge on
divisions.head()

state     region     division
0    Connecticut  Northeast  New England
1          Maine  Northeast  New England
2  Massachusetts  Northeast  New England
3  New Hampshire  Northeast  New England
4   Rhode Island  Northeast  New England

In [31]:
# Merge the datasets on 'state'
census = populations.merge(divisions, on = 'state')

census.head()

state      2010      2011      2012      2013      2014      2015  \
0     Alabama   4785492   4799918   4815960   4829479   4843214   4853875   
1      Alaska    714031    722713    731089    736879    736705    737709   
2     Arizona   6408312   6467163   6549634   6624617   6719993   6817565   
3    Arkansas   2921995   2939493   2950685   2958663   2966912   2977853   
4  California  37332685  37676861  38011074  38335203  38680810  38993940   

       2016 region            division  
0   4863300  South  East South Central  
1    741894   West             Pacific  
2   6931071   West            Mountain  
3   2988248  South  West South Central  
4  39250017   West             Pacific

12- Re-shape census so that one column contains all population measures, and another the year attributes.


In [53]:
# Reshape the census data into 'long' format using pd.melt() function
census_long = pd.melt(census, id_vars = ['state','division','region'], var_name = 'Year', value_name = 'Population')
census_long.head()

state            division region  Year  Population
0     Alabama  East South Central  South  2010     4785492
1      Alaska             Pacific   West  2010      714031
2     Arizona            Mountain   West  2010     6408312
3    Arkansas  West South Central  South  2010     2921995
4  California             Pacific   West  2010    37332685

13- Group the data by year and summarize it.

In [54]:
# Convert 'Population' column to numeric type in order to carry out agg function
census_long['Population'] = pd.to_numeric(census_long['Population'])


In [55]:
# Summarise the data using .agg() function and some standard statistical summaries
summary = census_long.groupby('Year').agg(total_pop = ('Population', 'sum'),
                                mean_pop = ('Population','mean'),
                                std_pop = ('Population', 'std'),
                                max_pop = ('Population', 'max'),
                                min_pop = ('Population','min'))
print(summary)

      total_pop      mean_pop       std_pop   max_pop  min_pop
Year                                                          
2010  309348193  6.065651e+06  6.839536e+06  37332685   564513
2011  311663358  6.111046e+06  6.907730e+06  37676861   567725
2012  313998379  6.156831e+06  6.974942e+06  38011074   576765
2013  316204908  6.200096e+06  7.038774e+06  38335203   582684
2014  318563456  6.246342e+06  7.110811e+06  38680810   583642
2015  320896618  6.292091e+06  7.181603e+06  38993940   586555
2016  323127513  6.335834e+06  7.243849e+06  39250017   585501


14- Group the data by region, division and year and summarize it.

In [58]:
summary2 = census_long.groupby(['region', 'division','Year']).agg(total_pop = ('Population', 'sum'),
                                mean_pop = ('Population','mean'),
                                std_pop = ('Population', 'std'),
                                max_pop = ('Population', 'max'),
                                min_pop = ('Population','min'))
print(summary2)

                                 total_pop    mean_pop       std_pop  \
region  division           Year                                        
Midwest East North Central 2010   46440847   9288169.4  3.115322e+06   
                           2011   46507169   9301433.8  3.109807e+06   
                           2012   46572795   9314559.0  3.104953e+06   
                           2013   46660465   9332093.0  3.099620e+06   
                           2014   46731329   9346265.8  3.091076e+06   
...                                    ...         ...           ...   
West    Pacific            2012   50928325  10185665.0  1.574263e+07   
                           2013   51372320  10274464.0  1.587648e+07   
                           2014   51856431  10371286.2  1.601909e+07   
                           2015   52341730  10468346.0  1.614496e+07   
                           2016   52801933  10560386.6  1.624360e+07   

                                  max_pop  min_pop  
region  di

15- Create a crosstab between state and category showing amount pledged as the value.
Please refer to the following link for this practice:

In [60]:
# Load the data
data = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv')
data.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [62]:
crosstab = pd.crosstab(data['state'], data['category'], values=data['usd_pledged'], aggfunc='sum')
print(crosstab)

category     3D Printing      Academic   Accessories         Action  \
state                                                                 
canceled    9.926480e+04    4152.45800  1.219638e+05   13413.619000   
failed      3.182202e+05   79895.73553  2.617623e+05   81820.666880   
live        4.540540e+03      12.51820  7.904410e+03       0.752403   
successful  8.095148e+06  345302.88929  1.995300e+06  218053.552000   
suspended   1.157608e+04     253.30890  3.773542e+03            NaN   
undefined   0.000000e+00           NaN  0.000000e+00            NaN   

category         Animals     Animation   Anthologies       Apparel  \
state                                                                
canceled        0.000000  5.743998e+04    4581.86113  9.038852e+04   
failed      14025.818316  4.655484e+05   15346.22450  3.604085e+05   
live                 NaN  4.937077e+02     572.08450  5.139040e+04   
successful  93671.235000  2.490226e+06  767299.12000  5.380367e+06   
suspended  